src/AutoVizuA11y.tsx

```diff
@@ -96,141 +106,111 @@ const AutoVizuA11y = ({
+	let alertDiv: React.ReactNode = useMemo(
+		() => (
+			<div
+				ref={alertDivRef}
+				className="a11y_alert visually-hidden"
+				role="alert"
+				aria-live="assertive"
+			>
+				{"\u00A0"}
+			</div>
+		),
+		[],
+	);
 
-	if (autoDescriptions) {
-		apiKey = autoDescriptions.apiKey;
-		model = autoDescriptions.model;
-		temperature = autoDescriptions.temperature;
-	}
-
-	if (insights == undefined || !(insights in data[0])) {
-		insights = "";
-	}
-
-	// Generate a unique identifier for this component
-	const componentId = newId();
-
-	const storedLongerKey = `oldLonger_${componentId}`;
-	const storedSmallerKey = `oldSmaller_${componentId}`;
-
-	const handleFocus = () => {
-		//css
-		ref.current!.classList.add("focused");
-		toolTutorial = localStorage.getItem("toolTutorial");
-		if (toolTutorial === "true") {
-			alertDiv = ref.current!.getElementsByClassName("a11y_alert")[0] as HTMLElement | null;
-			if (alertDiv) {
-				alertDiv.textContent =
-					"You just entered an AutoVizually chart." +
-					" For information on how to interact with it, press the question mark key" +
-					" to open the shortcut guide";
+	let chartDescription: React.ReactNode = useMemo(
+		() => (
+			<p
+				style={{ textIndent: "-10000px" }}
+				className="a11y_desc visually-hidden"
+				data-testid="a11y_desc"
+				onFocus={() => {
+					handleFirstFocus({ alertDiv, chartRef, alertDivRef });
+				}}
+				onBlur={(_) => handleBlur(chartRef)}
+				aria-label={descriptionContent}
+			>
+				{descriptionContent}
+			</p>
+		),
+		[descriptionContent],
+	);
+
+	let chart = useMemo(
+		() => React.Children.map(children, (child) => <div>{child}</div>),
+		[children],
+	);
+
+	useEffect(() => {
+		if (chartRef.current) {
+			let newElements: HTMLElement[] = [];
+			if (selectorType.element !== undefined) {
+				newElements = Array.from(chartRef.current.querySelectorAll(selectorType.element));
+			} else {
+				newElements = Array.from(
+					chartRef.current.getElementsByClassName(selectorType.className || ""),
+				) as HTMLElement[];
 			}
-			setTimeout(function () {
-				if (alertDiv) {
-					alertDiv.textContent = "\u00A0";
-				}
-			}, 1000);
-			localStorage.setItem("toolTutorial", "false");
-			toolTutorial = "false";
+			setElements(newElements);
 		}
-	};
+	}, [chartRef, selectorType]);
 
-	const handleBlur = () => {
-		ref.current!.classList.remove("focused");
-	};
+	useEffect(() => {
+		// features exclusive to bar charts (might be able to turn this more modular)
+		if (!selectorType) {
+			console.log("Type of chart not supported or no type given");
+		}
 
-	// Function to add an object to the array
-	const nextSeries = () => {
-		let currentPos = series.indexOf(selectedSeries);
-		let nextPos = (currentPos + 1) % series.length;
-		setSelectedSeries(series[nextPos]);
-	};
+		// verify if an insights key was passed and it exists in the data
+		if (insights == undefined || !(insights in data[0])) {
+			insights = "";
+		}
+	}, []);
 
 	useEffect(() => {
```

Comment:
This is a complex useEffect, so we might need to break it down into smaller chunks and change some of the logic. Here's my suggestion.

```suggestion
import { wait, getLSItem, setLSItem } from "@feedzai/js-utilities";

// Keep this const outside of the component code
const TIME_TO_WAIT_BEFORE_HANDLING_DESCRIPTIONS = 500;

// Outside the component scope
function initToolTutorial() {
  let toolTutorial = getLSItem("toolTutorial");
 
  if (!toolTutorial) {
    setLSItem("toolTutorial", "true");
  }
}; 

....

useEffect(async () => {
  if (!chartRef.current) return;

  const initSeries = () => {
    if (multiSeries) {
      const uniqueValues = [...new Set(data.map((item: any) => item[multiSeries]))];
      setSeries(uniqueValues);
      setSelectedSeries(uniqueValues[0]);
    }
  };

  const handleDescriptions = async () => {
    const storedLongerKey = `oldLonger_${componentId}`;
    const storedSmallerKey = `oldSmaller_${componentId}`;
    
    const result = await arrayConverter(data, insights);
    setArrayConverted(result);

    let insightsArrayAux: number[] = [];
    let averageAux = 0;
    if (insights !== "") {
      insightsArrayAux = insightsCalculator(result);
      averageAux = insightsArrayAux[1];
    }

    addAriaLabels({ chartRef, descriptor, selectorType, data, multiSeries });

    let chartDescriptions: string[] = [];
    if (autoDescriptions?.dynamicDescriptions === false) {
      chartDescriptions = [
        getLSItem(storedLongerKey)!,
        getLSItem(storedSmallerKey)!,
      ];
    } else if (manualDescriptions) {
      chartDescriptions = [manualDescriptions.longer, manualDescriptions.shorter];
    }

    if (chartDescriptions[0] && chartDescriptions[1]) {
      setDescs(chartDescriptions);
      descriptionsKeyHandler({
        chartRef,
        setDescriptionContent,
        type,
        descs: chartDescriptions,
        title,
        autoDescriptions,
      });
    } else {
      const generatedDescriptions = await generateDescriptions({
        title,
        data,
        average: averageAux,
        context,
        apiKey: autoDescriptions!.apiKey,
        model: autoDescriptions!.model,
        temperature: autoDescriptions!.temperature,
      });
      setDescs(generatedDescriptions);
      descriptionsKeyHandler({
        chartRef,
        setDescriptionContent,
        type,
        descs: generatedDescriptions,
        title,
        autoDescriptions,
      });
      if (autoDescriptions?.dynamicDescriptions === false) {
        setLSItem(storedLongerKey, generatedDescriptions[0]);
        setLSItem(storedSmallerKey, generatedDescriptions[1]);
      }
    }
  };

  initToolTutorial();
  initSeries();
  
  const timer = await wait(TIME_TO_WAIT_BEFORE_HANDLING_DESCRIPTIONS);

  handleDescriptions();
  switchToChartLevel(chartRef, true);

  // Cleanup function to clear the timeout if the component unmounts
  return () => clearTimeout(timer);

}, [chartRef, data, insights, multiSeries, componentId, autoDescriptions, manualDescriptions, title, type]);
```

This is a rough implementation of the same thing using async/await and moving some of the functions outside of the component scope. Test it out, some things might not work out of the box, it was just a simple refactor in a few minutes.

Still, it seems that there's too much going on for someone to be able to read and parse this all at once.
So, my suggestion would be:
- Try to follow this code cleanup approach, moving some things out of the component scope and also using async/await syntax to make it more readable.
- Then, see if you can take more things out of the component scope. Perhaps this needs to be a set of different, smaller functions that call each other, with as less side effects as possible, in the way of functional programming.

The objective is to make it do the same, but in a more readable fashion. To me, this seems like an information overload.

Reply:
I just found out that there's no documentation on the website about the storage helpers that we have on js-utilities and that I used here in the suggested code.

But you can find them out at:
https://github.com/feedzai/js-utilities/blob/main/src/functions/storage/local-storage.ts

Reply:
Tried to cleanup this useffect as much as possible. Had to change your approach quite a bit since `useEffect(async () => { ...` is not possible, apparently. Yet, I transformed the logic of the descriptions' generation into the suggested async/await one.